sync/nodegit.js

```diff
@@ -0,0 +1,136 @@
+/**
+ * Copyright(c) cnpm and other contributors.
+ * MIT Licensed
+ *
+ * Authors:
+ *   fengmk2 <fengmk2@gmail.com> (http://fengmk2.com)
+ */
+
+'use strict';
+
+/**
+ * Module dependencies.
+ */
+
+var debug = require('debug')('mirrors:sync:nodegit');
+var util = require('util');
+var urllib = require('urllib');
+var Syncer = require('./syncer');
+
+module.exports = NodeGitSyncer;
+
+function NodeGitSyncer(options) {
+  if (!(this instanceof NodeGitSyncer)) {
+    return new NodeGitSyncer(options);
+  }
+  Syncer.call(this, options);
+  this._npmPackageUrl = 'https://registry.npmjs.com/nodegit';
+  this._storeUrl = 'https://nodegit.s3.amazonaws.com/nodegit/nodegit/';
+}
+
+util.inherits(NodeGitSyncer, Syncer);
+
+var proto = NodeGitSyncer.prototype;
+
+proto.check = function() {
+  return true;
+};
+
+proto.listdiff = function* (fullname, dirIndex) {
+  if (dirIndex !== 0) {
+    return [];
+  }
+
+  var existsCount = 0;
+  var existDirs = yield this.listExists('/');
+  var existDirsMap = {};
+  for (var i = 0; i < existDirs.length; i++) {
+    var item = existDirs[i];
+    existDirsMap[item.name] = true;
+  }
+  var result = yield urllib.request(this._npmPackageUrl, {
+    timeout: 60000,
+    dataType: 'json',
+    gzip: true,
+    followRedirect: true,
+  });
+  var versions = result.data.versions || {};
+  var needs = [];
+  for (var version in versions) {
+    var pkg = versions[version];
+    var binaryInfo = pkg.binary || {};
+    pkg.dirname = 'v' + pkg.version + '/';
+    pkg.publish_time = result.data.time[pkg.version];
+    if (existDirsMap[pkg.dirname]) {
+      existsCount++;
+      continue;
+    }
+    if (binaryInfo.host && binaryInfo.host.indexOf(this._storeUrl) === 0) {
```

Comment:
嗯，没法同步

Sent from my iPhone

> On Apr 14, 2016, at 4:53 PM, Yiyu He <notifications@github.com> wrote:
> 
> In sync/nodegit.js:
> 
> > +    dataType: 'json',
> > +    gzip: true,
> > +    followRedirect: true,
> > +  });
> > +  var versions = result.data.versions || {};
> > +  var needs = [];
> > +  for (var version in versions) {
> > +    var pkg = versions[version];
> > +    var binaryInfo = pkg.binary || {};
> > +    pkg.dirname = 'v' + pkg.version + '/';
> > +    pkg.publish_time = result.data.time[pkg.version];
> > +    if (existDirsMap[pkg.dirname]) {
> > +      existsCount++;
> > +      continue;
> > +    }
> > +    if (binaryInfo.host && binaryInfo.host.indexOf(this._storeUrl) === 0) {
> 前期有一些老版本不是用这个 host 的就不管了？
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
>